keccak256/src/circuit/padding.rs

```diff
@@ -0,0 +1,443 @@
+        let q_all = meta.selector();
+        let q_without_first = meta.selector();
+        let q_without_last = meta.selector();
+        let q_last = meta.selector();
+        let is_finalize = meta.advice_column();
+        let byte = meta.advice_column();
+        let input_len = meta.advice_column();
+        let acc_len = meta.advice_column();
+        let diff_inv = meta.advice_column();
+        let is_pad_zone = meta.advice_column();
+        let padded_byte = meta.advice_column();
+        let word = meta.advice_column();
+        meta.enable_equality(is_finalize);
+        meta.enable_equality(input_len);
+        meta.enable_equality(acc_len);
+        let one = Expression::Constant(F::one());
+        let diff_is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(q_all),
+            |meta| {
+                meta.query_advice(input_len, Rotation::cur())
+                    - meta.query_advice(acc_len, Rotation::cur())
+            },
+            diff_inv,
+        );
+        let word_config = WordConfig::configure(meta, padded_byte, word);
+        meta.create_gate("all", |meta| {
+            let q_all = meta.query_selector(q_all);
+            let is_pad_zone_cur = meta.query_advice(is_pad_zone, Rotation::cur());
+            let byte_cur = meta.query_advice(byte, Rotation::cur());
+
+            vec![q_all * (is_pad_zone_cur * byte_cur)]
+        });
+
+        meta.create_gate("without last", |meta| {
+            let q_without_last = meta.query_selector(q_without_last);
+            let acc_len_cur = meta.query_advice(acc_len, Rotation::cur());
+            let acc_len_next = meta.query_advice(acc_len, Rotation::next());
+            let padded_byte_cur = meta.query_advice(padded_byte, Rotation::cur());
+            let byte_cur = meta.query_advice(byte, Rotation::cur());
+            iter::empty()
+                .chain(Some(("increase acc_len", acc_len_next - acc_len_cur - one)))
+                .chain(Some((
+                    "check padded byte",
+                    padded_byte_cur
+                        - byte_cur
+                        - diff_is_zero.clone().is_zero_expression
+                            * Expression::Constant(F::from(0x80)),
+                )))
+                .map(move |(name, poly)| (name, q_without_last.clone() * poly))
+        });
+        meta.create_gate("without first", |meta| {
+            let q_without_first = meta.query_selector(q_without_first);
+            let is_pad_zone_prev = meta.query_advice(is_pad_zone, Rotation::prev());
+            let is_pad_zone_cur = meta.query_advice(is_pad_zone, Rotation::cur());
+            vec![(
+                "check pad_zone",
+                q_without_first
+                    * (is_pad_zone_cur
+                        - is_pad_zone_prev
+                        - diff_is_zero.clone().is_zero_expression),
+            )]
+        });
```

Comment:
Makes sense! I think that maybe adding a really short description on the top of each gate would be nice. Since `without_first` doesn't really showcase anything related to what is done in this gate.